### ISSUE TITLE
3w: Bulk upload existing data

### ISSUE THREAD

**Issue by frozenhelium** [IFRCGo/go-api#531]
Create a bulk upload for existing 3w project data. We can use admin panel to upload a `csv` file with predefined template (only feasible if the size of csv file is not too large). Or, If its a one time thing, we can create fixture to load the data.

**Comment by frozenhelium** [comment]
@batpad What do you think?

**Comment by batpad** [comment]
@frozenhelium :+1: I think the expectation here is that this would be something users can do themselves. We should figure out / set expectations - I imagine this is something only admin users would be able to do and we probably need to have some documentation about being careful with this feature. I imagine a CSV upload in the admin should be good, yes. It's probably completely fine to enforce a max size of the CSV file.

**Comment by frozenhelium** [comment]
> @frozenhelium  I think the expectation here is that this would be something users can do themselves. We should figure out / set expectations - I imagine this is something only admin users would be able to do and we probably need to have some documentation about being careful with this feature. I imagine a CSV upload in the admin should be good, yes. It's probably completely fine to enforce a max size of the CSV file.

Sounds good. We'll go with the admin panel options then.

**Comment by batpad** [comment]
@frozenhelium it maay be useful to explore how easy / hard this would be to do just in the React app and not in the Django admin -- I know there is a general desire to move away from the Django Admin for as much as possible.

I also think there's some other places in the app that might benefit from having a good way to upload files to the backend. Is it possible to scope out both options and give an estimate of how much work you think each one is? I.e. uploading a CSV as part of the frontend app, and implementing as a function in the Django admin? And that may help better take a decision on how to implement it?

**Comment by frozenhelium** [comment]
Sure, I'll discuss it and provide an estimate.

**Comment by batpad** [comment]
As per @LukeCaley if we need this, we need this only as a one-time operation to import existing CSV data into the database

**Comment by batpad** [comment]
@LukeCaley to share samples of data

**Comment by nanometrenat** [comment]
@LukeCaley is this still required?

**Comment by LukeCaley** [comment]
Yes, this will still be required.  Sample data from WASH, DRR, and other teams being compiled as part of the planning for launch (cc @lauripynnonen) of the feature.

**Comment by lauripynnonen** [comment]
@frozenhelium - basically we would need to be able to upload all the contents of the Add 3W Activity-form through a csv from the admin-backend. If we make the csv ourselves the error handling does not need to be massive I assume?

**Comment by thenav56** [comment]
@lauripynnonen - Sounds good with the admin-backend, Can you provide a sample CSV we can look into?.

**Comment by lauripynnonen** [comment]
@thenav56 Thanks Navin! You can get a sample csv by exporting from the 3w

**Comment by thenav56** [comment]
@lauripynnonen Thanks for the information about the sample CSV.
Can you let me know if we also need to track import activities, such as who imported? what/when was imported? import result?

**Comment by lauripynnonen** [comment]
@thenav56 that would be useful, meaning collecting timestamp, username, action (bulk upload), *import result

*if the log is saved to each activity, import result would only be saved to successfully uploaded activities, which would mean the import result would always be "success"?

**Comment by thenav56** [comment]
@lauripynnonen Yeah, we process all the project data within the CSV, if all are valid then only we save it to the database. In case of error, we only save/show the error.

**Comment by frozenhelium** [comment]
@lauripynnonen to import the projects: https://go-api.togglecorp.com/deployments/project/

**Comment by lauripynnonen** [comment]
I exported all projects from Bangladesh, changed the country to Afganistan and set them all to Countrywide (this is the file in txt [projects-export-1586872881784.txt](https://github.com/IFRCGo/go-api/files/4475780/projects-export-1586872881784.txt) - filetype restrictions in GitHub...), ended up with the following error 

> Importing projects-export-1586872881784.csv failed. Check file and try again!!
> Traceback (most recent call last):
>   File "/home/ifrc/go-api/deployments/forms.py", line 104, in handle_bulk_upload
>     projects = self._handle_bulk_upload(request.user, file)
>   File "/home/ifrc/go-api/deployments/forms.py", line 80, in _handle_bulk_upload
>     start_date=strip_date(row['Start Date']),
>   File "/home/ifrc/go-api/deployments/forms.py", line 46, in strip_date
>     return datetime.datetime.strptime(date, "%Y-%m-%d").date()
>   File "/usr/local/lib/python3.6/_strptime.py", line 565, in _strptime_datetime
>     tt, fraction = _strptime(data_string, format)
>   File "/usr/local/lib/python3.6/_strptime.py", line 362, in _strptime
>     (data_string, format))
> ValueError: time data '1/1/2020' does not match format '%Y-%m-%d'

So day-format-out is different from day-format-in

**Comment by lauripynnonen** [comment]
Aand I tested by adding a dummy number to all fields and it went through

**Comment by thenav56** [comment]
@lauripynnonen Thanks for the detail comment. I just updated the server.

Now the original CSV file should work fine. The spelling of Afghanistan is incorrect in the file, make sure to change it before upload.

Let me know if you have any more feedback.

**Comment by karitotp** [comment]
I tested also with a file of exported projects from Bangladesh, I changed the Country to Brazil and the Region to Countrywide, I did not get the previous errors, what I am getting now is errors due Primary Sector that are type of health (public) and health (clinical). Same happen when I want to import a file that have only one project that has Primary Sector as health (public) or health (clinical).

> Importing banglateest.csv failed. Check file and try again!!
Traceback (most recent call last):
  File "/home/ifrc/go-api/deployments/forms.py", line 127, in handle_bulk_upload
    projects = self._handle_bulk_upload(request.user, file)
  File "/home/ifrc/go-api/deployments/forms.py", line 96, in _handle_bulk_upload
    primary_sector=sectors[key_clean(row['Primary Sector'])],
KeyError: 'health (public)'

>Importing import test.csv failed. Check file and try again!!
Traceback (most recent call last):
  File "/home/ifrc/go-api/deployments/forms.py", line 127, in handle_bulk_upload
    projects = self._handle_bulk_upload(request.user, file)
  File "/home/ifrc/go-api/deployments/forms.py", line 96, in _handle_bulk_upload
    primary_sector=sectors[key_clean(row['Primary Sector'])],
KeyError: 'health (clinical)'

I changed the value of Primary Sector for another different to health (public) and health (clinical), and it was imported correctly.
